src/main/java/me/desht/modularrouters/item/module/ITargetedModule.java

```diff
@@ -0,0 +1,117 @@
+     * <p>Note: it is not guaranteed that the module will only have blocks that pass this test selected, see {@link AddModuleTargetEvent}.
+     */
+    @ApiStatus.OverrideOnly
+    default boolean isValidTarget(UseOnContext ctx) {
+        return InventoryUtils.getInventory(ctx.getLevel(), ctx.getClickedPos(), ctx.getClickedFace()).isPresent();
+    }
+
+    default int getMaxTargets() {
+        return 1;
+    }
+
+    /**
+     * Does this module have limited range?
+     *
+     * @return true if range is limited, false otherwise
+     */
+    default boolean isRangeLimited() {
+        return true;
+    }
+
+    /**
+     * {@return whether this module can operate in the given {@code dimension}}
+     */
+    default boolean canOperateInDimension(ResourceKey<Level> dimension) {
+        return true;
+    }
+
+    /**
+     * Retrieve multi-targeting information from a module itemstack.
+     *
+     * @param stack the module item stack
+     * @param checkBlockName verify the name of the target block - only works server-side
+     * @implNote the set will be limited to the max target amount of the module
+     * @return a set of targets for the module
+     */
+    static Set<ModuleTarget> getTargets(ItemStack stack, boolean checkBlockName) {
+        var max = ((ITargetedModule) stack.getItem()).getMaxTargets();
+        Set<ModuleTarget> result = Sets.newHashSet();
+
+        boolean update = false;
+        var targets = stack.getOrDefault(ModDataComponents.MODULE_TARGET_LIST, ModuleTargetList.EMPTY).targets();
+        for (int i = 0; i < targets.size() && result.size() < max; i++) {
+            var target = targets.get(i);
+            if (checkBlockName) {
+                var newTarget = updateTargetBlockName(stack, target);
+                if (newTarget != target) update = true;
+                target = newTarget;
+            }
+            if (target != null) {
+                result.add(target);
+            }
+        }
+
+        if (update) {
+            stack.set(ModDataComponents.MODULE_TARGET_LIST, new ModuleTargetList(List.copyOf(result)));
+        }
+
+        return result;
+    }
+
+    /**
+     * Sets the targets of a module.
+     * @param stack the module stack to update
+     * @param targets the new targets
+     */
+    static void setTargets(ItemStack stack, Collection<ModuleTarget> targets) {
+        stack.set(ModDataComponents.MODULE_TARGET_LIST, new ModuleTargetList(List.copyOf(targets)));
+    }
+
+    /**
+     * Checks if the module can select the target of the {@code context}.
+     */
+    static boolean canSelectTarget(UseOnContext context) {
+        var module = context.getItemInHand().getItem();
+        return NeoForge.EVENT_BUS.post(new AddModuleTargetEvent((ModuleItem) module, context, ((ITargetedModule) module).isValidTarget(context))).isValid();
+    }
+
+    private static ModuleTarget updateTargetBlockName(ItemStack stack, ModuleTarget target) {
```

Comment:
`stack` isn't actually used in this method...

Reply:
Ah, I forgot to remove it. There's no need for it as the component is updated in `getTargets` now. Removed it